dimod/generators/bpsp.py

```diff
@@ -0,0 +1,94 @@
+# Copyright 2024 D-Wave Systems Inc.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+__all__ = ["random_binary_paint_shop_problem"]
+
+import typing
+import numpy.random
+
+import dimod
+
+
+def binary_paint_shop_problem(car_sequence: typing.Sequence[typing.Hashable]) -> dimod.BinaryQuadraticModel:
+    """Generates a binary quadratic model representing a binary paint shop problem.
+
+    Ising formulation for Binary Paint Shop Problem:
+    - Streif, Michael, et al.
+      "Beating classical heuristics for the binary paint shop
+      problem with the quantum approximate optimization algorithm."
+      Physical Review A 104.1 (2021): 012403.
+
+    Args:
+        car_sequence: A list of car labels.
+
+    Returns:
+        Ising model where each variable is a car, and the value indicates what color it gets painted first.
+    """
+
+    car_counter = {car: 0 for car in car_sequence}
+    if len(car_sequence) != 2 * len(car_counter):
+        raise ValueError("Car labels are not unique")
+
+    bqm = dimod.BinaryQuadraticModel(vartype=dimod.SPIN)
+    for car1, car2 in zip(car_sequence, car_sequence[1:]):
+        if car1 != car2:
+            bqm.add_quadratic(car1, car2, (-1) ** (car_counter[car1] + car_counter[car2] + 1))
+        car_counter[car1] += 1
+    return bqm
+
+
+def random_binary_paint_shop_problem(n_cars: int, seed: int = None) -> dimod.BinaryQuadraticModel:
+    """Generates a random binary paint shop problem.
+
+    Binary Paint Shop Problem:
+    - Epping, Th, Winfried Hochstättler, and Peter Oertel.
+      "Complexity results on a paint shop problem."
+      Discrete Applied Mathematics 136.2-3 (2004): 217-226.
+
+    Args:
+        n_cars: Number of cars.
+        seed: Seed for the random number generator.
+
+    Returns:
+        Ising model for the random binary paint shop problem.
+    """
+
+    rng = numpy.random.RandomState(seed)
```

Comment:
`RandomState` is [legacy](https://numpy.org/doc/stable/reference/random/legacy.html) and likely to be deprecated at some point. So we should use [default_rng()](https://numpy.org/doc/stable/reference/random/generator.html) instead.